eos-paygd/main.c

```diff
@@ -35,8 +35,90 @@
 #include <sys/un.h>
 #include <libeos-payg/efi.h>
 
+#define LOGFILE_DIRNAME "/var/log/eos-payg"
+#define LOGFILE_BASENAME "eos-paygd"
+#define LOGFILE_EXT "log"
+
 #define TIMEOUT_POWEROFF_ON_ERROR_MINUTES 20
 
+static GIOChannel *
+open_log_file (void)
+{
+  const gchar *log_file_name = NULL;
+  g_autoptr(GFile) log_file = NULL;
+  g_autoptr(GDateTime) now = NULL;
+  g_autofree gchar *tstamp = NULL;
+
+  /* Build log file name with a date stamp so we get one file per day,
+   * ex., eos-paygd-20220418.log
+   */
+  now = g_date_time_new_now_local ();
+  tstamp = g_date_time_format (now, "%Y%m%d");
+  log_file_name = g_strconcat (LOGFILE_DIRNAME, "/", LOGFILE_BASENAME, "-",
+                               tstamp, ".", LOGFILE_EXT, NULL);
+  log_file = g_file_new_for_path (log_file_name);
+
+  /* We can't log an error here if this fails, as it would cause infinite
+   * recursion */
+  return g_io_channel_new_file (log_file_name, "a", NULL);
+}
+
+/* we can't use g_log_writer_default_would_drop until glib 2.68, which will be
+ * available on EOS 5.0+, so we are copying a simplified version of its
+ * implementation here.
+ */
+#define DEFAULT_LEVELS (G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE)
+#define INFO_LEVELS (G_LOG_LEVEL_INFO | G_LOG_LEVEL_DEBUG)
+static gboolean
+should_drop_message (GLogLevelFlags log_level)
+{
+  /* Disable debug message output unless specified in G_MESSAGES_DEBUG. */
+  if (!(log_level & DEFAULT_LEVELS) && !(log_level >> G_LOG_LEVEL_USER_SHIFT))
+    {
+      const gchar *domains = g_getenv ("G_MESSAGES_DEBUG");
+
+      if ((log_level & INFO_LEVELS) == 0 ||
+          domains == NULL)
+        return TRUE;
+
+      if (strcmp (domains, "all") != 0)
+        return TRUE;
+    }
+
+  return FALSE;
+}
+
+/* Custom log writer function that saves messages to a separate file before
+ * forwarding them to the default writer function.
+ */
+static GLogWriterOutput
+log_writer (GLogLevelFlags log_level,
+            const GLogField* fields,
+            gsize n_fields,
+            gpointer user_data)
+{
+  GIOChannel *log_io = NULL;
+
+  g_return_val_if_fail (fields != NULL, G_LOG_WRITER_UNHANDLED);
+  g_return_val_if_fail (n_fields > 0, G_LOG_WRITER_UNHANDLED);
+
+  if (should_drop_message (log_level))
+    return G_LOG_WRITER_HANDLED;
+
+  log_io = open_log_file ();
```

Comment:
Should this be opening the log file, writing to it, and closing it again every time a message is logged? That seems costly.

Would it be better to keep a persistent handle to the open log file?

Reply:
Re-opening the file every time is the way we designed cheap log rotation: we can't use `logrotate` because eos-payg can't receive signals in "phase-4 mode", and while there are other design alternatives (use logrotate and monitor the fd for renames with inotify, monitor the date or file size and close + re-open at that point), they are all more complicated and this seems good enough for what we need while remaining relatively pretty simple.

This designed choice is mentioned in the commit message of https://github.com/endlessm/eos-payg/commit/4a98f292cce82d260423aef614f7aede580a58e1, but it seems like we should also add it as a comment in the code.

Reply:
:+1: